common/canonical_serialization/src/simple_serializer.rs

```diff
@@ -58,29 +58,33 @@ where
         Ok(self)
     }
 
-    fn encode_btreemap<K: CanonicalSerialize, V: CanonicalSerialize>(
+    fn encode_tuple_iterator<K: CanonicalSerialize, V: CanonicalSerialize, I>(
         &mut self,
-        v: &BTreeMap<K, V>,
-    ) -> Result<&mut Self> {
+        iter: I,
+    ) -> Result<&mut Self>
+    where
+        I: Iterator<Item = (K, V)>,
+    {
+        let mut map = BTreeMap::new();
+        for (key, value) in iter {
+            map.insert(
+                SimpleSerializer::<Vec<u8>>::serialize(&key)?,
+                SimpleSerializer::<Vec<u8>>::serialize(&value)?,
+            );
+        }
+
         ensure!(
```

Comment:
Are we changing the semantics here? Should I add the length check at each round of iteration?

Reply:
No, this is purely the number of elements and not the total space occupied by the map.